### ISSUE TITLE
Optionally validate entire schema before throwing exception

### ISSUE THREAD

**Issue by alecwcp** [thephpleague/openapi-psr7-validator#35]
Currently when an error occurs (mismatch between data and schema) validation stops at the first error.

It would be useful to carry on validating the data to return as many errors at once as possible (e.g. so a consumer can be sent the errors from their request and they can address them all at once, or we can validate our response as it is sent and log any mismatches with our schema).

The snippet below shows the issue, the exception only includes details of the first issue (type error for `title` field) but doesn't include details of the second error (keyword mismatch due to the presence of `dummy` field, additional properties are not allowed).

```
use cebe\openapi\Reader as OpenApiReader;
use cebe\openapi\spec\Schema as OpenApiSchema;
use League\OpenAPIValidation\Schema\SchemaValidator;

$spec = <<<SPEC
schema:
  type: object
  additionalProperties: false
  required:
    - title
  properties:
    title:
      type: string
    aNumber:
      type: integer
      nullable: true
SPEC;

$spec = OpenApiReader::readFromYaml($spec);
$schema = new OpenApiSchema($spec->schema);

$data = [
    'title' => 23,
    'dummy' => 'blah'
];

try {
    (new SchemaValidator(SchemaValidator::VALIDATE_AS_REQUEST))->validate($data, $schema);
} catch(\League\OpenAPIValidation\Schema\Exception\KeywordMismatch $e) {
    // you can evaluate failure details
    // $e->keyword() == "enum"
    // $e->data() == "c"
    // $e->dataBreadCrumb()->buildChain() -- only for nested data
    var_dump($e->keyword());
    var_dump($e->data());
    var_dump($e->dataBreadCrumb());
    throw $e;
}
```

I realise that this would be quite a backwards compatibility break if this functionality always happened, but it could possibly be set as an option on the `SchemaValidator`?

If it's agreed that this would be useful functionality to have then I'm happy to work on a PR for it.

**Comment by scaytrase** [comment]
The problem of full validation is that some of the validators rely on previous validation succeeded. I.e you have schema
```YAML
schema:
  type: object
  additionalProperties: false
  required:
    - strings
  properties:
    strings:
      type: array
      minItems: 10
      items:
         type: string
```

If you pass there a string instead of array any further validations have no-sense and will fail, because of invalid data passed to validator. I.e you can't validate `minItems` of `string`.

So entire schema validation is not possible.

But I think we can gather more failed validations as option, yes. But I think this should be option for `validate` method instead (i.e `fail-fast` mode by default, and `complete` optional).

**Comment by alecwcp** [comment]
Yes, you're right. Depending on the actually data submitted it is highly unlikely the entire schema can be validated, I should have said "validate as much as possible".

An option passed to `validate` that is one of `fail-fast` (default) or `complete` is a good idea :+1: 

For getting a list of the errors we could return an array from the `validate` method, however that would mean we're changing the interface https://github.com/thephpleague/openapi-psr7-validator/blob/master/src/Schema/Validator.php - what's your opinion on this?
The other option would seem to be to throw a new type of exception that includes a list of errors, each of which has links to any nested errors.

I've found `Symfony\Component\Validator\ConstraintViolationListInterface` and `Symfony\Component\Validator\ConstraintViolationInterface` nice to work with previously - any thoughts on using those for the returned errors when we're returning as many as possible? Not sure if you want to bring that in as a dependency of this project.

**Comment by scaytrase** [comment]
I think we can try to adopt `symfony/validator` interfaces and see how it looks in your PR. @lezhnev74 WDYT?
I hope borrowing some logic from another specified lib for validation could make this one a bit easier. `ConstraintViolation` has already whatever we want - breadcrumbs, tempated messages, useful data in violation. And many project including mine are already hanling this violations for API stuff (like entity validation), so it would be easy to adopt on user side

**Comment by lezhnev74** [comment]
From the top of my head:
- we have a growing list of optional features (cache, validation etc). It looks like the lib needs to be extendable (new functionality should be added as plugins rather than changing core components each time we need some new optional thing). I would envision the architecture with hooks/events where you can inject extensions.
- Symfony's ecosystem of packages is well known and I like the idea of reusing their packages here. Does it have any unwanted dependencies for us?

**Comment by scaytrase** [comment]
> list of optional features (cache, validation etc)

I hope validation would never be _optional_ for this lib 😂 But I got your idea, yea, we need to think in this direction.

As for dependencies - `symfony/validator` has [pretty simple requirements](https://github.com/symfony/validator/blob/master/composer.json#L18-L23), so I think we have no problems with it.

**Comment by lezhnev74** [comment]
Yeah I meant the greedy validation which won't stop on first error. :)

**Comment by scaytrase** [comment]
> If you pass there a string instead of array any further validations have no-sense and will fail, because of invalid data passed to validator. I.e you can't validate minItems of string.
>
> So entire schema validation is not possible.

btw to keep things up. According to discussions and links from
https://github.com/thephpleague/openapi-psr7-validator/issues/32#issuecomment-552660332
it's fully valid to validate a `string` with `minItems: 10`. Validator should pass if the datatype is not countable

**Comment by iosifch** [comment]
Any news about this feature? How can I help?
